sql/000002.sql

```diff
@@ -0,0 +1,176 @@
+alter table graphile_worker.jobs add column key text unique check(length(key) > 0);
+
+alter table graphile_worker.jobs add locked_at timestamptz;
+alter table graphile_worker.jobs add locked_by text;
+
+-- update any in-flight jobs
+update graphile_worker.jobs
+  set locked_at = q.locked_at, locked_by = q.locked_by
+  from graphile_worker.job_queues q
+  where q.queue_name = jobs.queue_name
+  and q.locked_at is not null;
+
+-- update add_job behaviour to meet new requirements
+drop function if exists graphile_worker.add_job(identifier text,
+  payload json,
+  queue_name text,
+  run_at timestamptz,
+  max_attempts int
+);
+create function graphile_worker.add_job(
+  identifier text,
+  payload json = '{}',
+  queue_name text = null,
+  run_at timestamptz = now(),
+  max_attempts int = 25,
+  job_key text = null
+) returns graphile_worker.jobs as $$
+declare
+  v_job graphile_worker.jobs;
+begin
+  if job_key is null then
+    insert into graphile_worker.jobs(task_identifier, payload, queue_name, run_at, max_attempts)
+      values(
+        identifier,
+        payload,
+        coalesce(queue_name, public.gen_random_uuid()::text),
+        run_at,
+        max_attempts
+      )
+      returning *
+      into v_job;
+  else
+    insert into graphile_worker.jobs (task_identifier, payload, queue_name, run_at, max_attempts, key)
+      values(
+        identifier,
+        payload,
+        coalesce(queue_name, public.gen_random_uuid()::text),
+        run_at,
+        max_attempts,
+        job_key
+      )
+      on conflict (key) do update set
+        -- update all job details other than queue_name, which we want to keep
+        -- the same unless explicitly provided
+        task_identifier=identifier,
```

Comment:
For this and the others, please reference the [`excluded` record](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT), rather than the arguments directly (which can in some cases share the name with the columns and make it less clear which values are being used).

```suggestion
        task_identifier = excluded.task_identifier,
```